### PR TITLE
feat(createAsyncThunk): async condition

### LINKED ISSUE #1494: Async `condition` in `createAsyncThunk`
I need my thunk to detect that another instance of the same operation is already running and to fulfill the promise (without dispatching any actions) only after that other instance is done. The code that dispatches the thunk doesn't care who started the operation, it just needs it to complete. As I have more than one such place in my code, I'd prefer to encapsulate this logic into the thunk, which would be possible if `options.condition` could return a promise. That looks like a simple change. I can give a shot at a PR.

### PR DESCRIPTION
**Pull request by thorn0**
closes #1494

### CODE REVIEW

File: packages/toolkit/src/createAsyncThunk.ts [hunk 1]
```diff
@@ -553,11 +553,15 @@ If you want to use the AbortController to react to \`abort\` events, please cons
       const promise = (async function () {
         let finalAction: ReturnType<typeof fulfilled | typeof rejected>
         try {
-          if (
-            options &&
-            options.condition &&
-            options.condition(arg, { getState, extra }) === false
-          ) {
+          let conditionResult = options?.condition?.(arg, { getState, extra })
+          if (conditionResult instanceof Promise) {
```
Comment (thorn0) [inline packages/toolkit/src/createAsyncThunk.ts:557]: Is `instanceof` good enough or do we want to support 3rd-party thenables?

Comment (thorn0) [reply packages/toolkit/src/createAsyncThunk.ts:557]: Also what should happen when an error is thrown in a synchronous `condition`?

Comment (phryneas) [reply packages/toolkit/src/createAsyncThunk.ts:557]: Hmm. `"then" in conditionResult" might be better, you're right.

As for an error in `condition`, so far that was just thrown as it was and I'm pretty okay. Might even make sense to remove the try..catch block above to mimick that.

Comment (thorn0) [reply packages/toolkit/src/createAsyncThunk.ts]: > As for an error in condition, so far that was just thrown as it was and I'm pretty okay. Might even make sense to remove the try..catch block above to mimick that.

Okay, let's do it that way. Bur I somewhat dislike the fact that in this case `rejected` is dispatched without corresponding `pending`.

**Review by phryneas — APPROVED** [review]

**Pull request merged**